File: ios/Classes/GyroProviderPlugin.swift
import Flutter

var _eventChannels: [String: FlutterEventChannel] = [:]
var _streamHandlers: [String: FlutterStreamHandler] = [:]
var _isCleanUp = false

public class GyroProviderPlugin: NSObject, FlutterPlugin {
    public static func register(with registrar: FlutterPluginRegistrar) {
        let gyroStreamHandler = GyroStreamHandler()
        let gyro = "com.gmail.wjlee611.gyro_event_channel"
        let gyroEventChannel = FlutterEventChannel(
            name: gyro,
            binaryMessenger: registrar.messenger()
        )
        gyroEventChannel.setStreamHandler(gyroStreamHandler)
        _eventChannels[gyro] = gyroEventChannel
        _streamHandlers[gyro] = gyroStreamHandler
        
        
        let rotateStreamHandler = RotateStreamHandler()
        let rotate = "com.gmail.wjlee611.rotate_event_channel"
        let rotateEventChannel = FlutterEventChannel(
            name: rotate,
            binaryMessenger: registrar.messenger()
        )
        rotateEventChannel.setStreamHandler(rotateStreamHandler)
        _eventChannels[rotate] = rotateEventChannel
        _streamHandlers[rotate] = rotateStreamHandler
        
        
        _isCleanUp = false
    }

    static func _cleanUp() {
        _isCleanUp = true
        for channel in _eventChannels.values {
            channel.setStreamHandler(nil)
        }
        _eventChannels.removeAll()
        for handler in _streamHandlers.values {
            handler.onCancel(withArguments: nil)
        }
        _streamHandlers.removeAll()
    }
}

File: android/src/main/kotlin/com/gmail/wjlee611/gyro_provider/GyroProviderPlugin.kt
package com.gmail.wjlee611.gyro_provider

import android.content.Context
import android.hardware.Sensor
import android.hardware.SensorManager

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel

/** GyroProviderPlugin */
class GyroProviderPlugin: FlutterPlugin {
  private lateinit var gyroEventChannel : EventChannel
  private lateinit var rotateEventChannel : EventChannel

  private lateinit var gyroStreamHandler : StreamHandlerImpl
  private lateinit var rotateStreamHandler : StreamHandlerImpl

  override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    setupEventChannel(flutterPluginBinding.applicationContext, flutterPluginBinding.binaryMessenger)
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    teardownEventChannel()
  }

  private fun setupEventChannel(context: Context, messenger: BinaryMessenger) {
    val sensorManager = context.getSystemService(Context.SENSOR_SERVICE) as SensorManager

    // GYROSCOPE
    gyroEventChannel = EventChannel(messenger, "com.gmail.wjlee611.gyro_event_channel")
    gyroStreamHandler = StreamHandlerImpl(sensorManager, Sensor.TYPE_GYROSCOPE)
    gyroEventChannel.setStreamHandler(gyroStreamHandler)

    // ROTATION
    rotateEventChannel = EventChannel(messenger, "com.gmail.wjlee611.rotate_event_channel")
    rotateStreamHandler = StreamHandlerImpl(sensorManager, Sensor.TYPE_GAME_ROTATION_VECTOR)
    rotateEventChannel.setStreamHandler(rotateStreamHandler)
  }

  private fun teardownEventChannel() {
    gyroEventChannel.setStreamHandler(null)
    gyroStreamHandler.onCancel(null)

    rotateEventChannel.setStreamHandler(null)
    rotateStreamHandler.onCancel(null)
  }
}
